src/Features/CSharp/Portable/CodeRefactorings/ConvertToRecord/CSharpConvertToRecordRefactoringProvider.cs

```diff
@@ -0,0 +1,623 @@
+                (type.TypeKind != TypeKind.Class && type.TypeKind != TypeKind.Struct) ||
+                // TODO: when adding other options it might make sense to convert a postional record to a non-positional one and vice versa
+                // but for now we don't convert records
+                type.IsRecord ||
+                // records can't be static and so if the class is static we probably shouldn't convert it
+                type.IsStatic ||
+                // make sure that there is at least one positional parameter we can introduce
+                !classDeclaration.Members.Any(m => ShouldConvertProperty(m, type)))
+            {
+                return;
+            }
+
+            var positionalTitle = FeaturesResources.Convert_to_positional_record;
+
+            var positional = CodeAction.Create(
+                positionalTitle,
+                cancellationToken => ConvertToParameterRecordAsync(document, type, classDeclaration, semanticModel, cancellationToken),
+                positionalTitle);
+            var nested = CodeAction.CodeActionWithNestedActions.Create(string.Format(FeaturesResources.Convert_0_to_record, type.Name), ImmutableArray.Create(positional), isInlinable: false);
+            context.RegisterRefactoring(nested);
+        }
+
+        private static async Task<Document> ConvertToParameterRecordAsync(
+            Document document,
+            INamedTypeSymbol originalType,
+            TypeDeclarationSyntax originalDeclarationNode,
+            SemanticModel semanticModel,
+            CancellationToken cancellationToken)
+        {
+            var codeGenerationService = document.GetRequiredLanguageService<ICodeGenerationService>();
+            var syntaxGenerator = document.GetRequiredLanguageService<SyntaxGenerator>();
+
+            var propertiesToMove = originalDeclarationNode.Members.Where(m => ShouldConvertProperty(m, originalType)).Cast<PropertyDeclarationSyntax>().AsImmutable();
+
+            var modifiedClassTrivia = GetModifiedClassTrivia(propertiesToMove, originalDeclarationNode);
+
+            var propertiesToAddAsParams = propertiesToMove.SelectAsArray(p =>
+                SyntaxFactory.Parameter(
+                    new SyntaxList<AttributeListSyntax>(p.AttributeLists.SelectAsArray(attributeList =>
+                        attributeList.Target == null
+                        // convert attributes attached to the property with no target into "property :" targeted attributes
+                        ? attributeList
+                            .WithTarget(SyntaxFactory.AttributeTargetSpecifier(SyntaxFactory.Token(SyntaxKind.PropertyKeyword)))
+                            .WithoutTrivia()
+                        : attributeList.WithoutTrivia())),
+                    modifiers: default,
+                    p.Type,
+                    p.Identifier,
+                    @default: null));
```

Comment:
note to self: are properties checked for bodies/initializers in ShouldConvertProperty?